libs/chain-signatures/contract/src/lib.rs

```
@@ -193,6 +279,45 @@ impl MpcContract {
         }
         Ok(())
     }
+
+    pub fn vote_code_hash(&mut self, code_hash: CodeHash) -> Result<(), Error> {
+        // TODO: Verify TEE quote here. See GitHub issue #378: https://github.com/Near-One/mpc/issues/378
+
+        if let ProtocolContractState::Running(state) = &self.protocol_state {
+            let participant = AuthenticatedParticipantId::new(state.parameters.participants())?;
+            let votes = self.tee_state.votes.vote(code_hash.clone(), &participant);
+
+            if votes >= self.threshold()?.value() {
+                self.tee_state.votes.clear_votes();
+                self.tee_state
+                    .historical_code_hashes
+                    .push(code_hash.clone());
+                self.tee_state.allowed_code_hashes.insert(AllowedCodeHash {
+                    code_hash,
+                    added: env::block_height(),
+                });
+                self.tee_state.upgrade_deadline = Some(env::block_height() + TEE_UPGRADE_PERIOD);
+            }
+        }
+
```

Lets return an error in case the contract is not in running state, since the vote is invalid in such a case.

---

Fixed in https://github.com/Near-One/mpc/pull/376/commits/cf0b0efeb3c17b9ed6e070456b40de88d77c2eae. ✅